docs:added docs for custom interaction (#20379)
newrelic/docs-website: src/content/docs/mobile-monitoring/new-relic-mobile-android/install-configure/android-agent-custom-Interaction.mdx

---

> title: Android Agent Custom Interactions

Reviewer: ```suggestion
title: Android agent custom interactions
```

tags:
  - Mobile monitoring
  - New Relic Mobile Android
  - Install configure
metaDescription: 'How to use Custom Interactions with Android Mobile Monitoring'
freshnessValidatedDate: 2025-04-02
---

<Callout variant="important">

>   This feature is only working correctly if you disable Default Interactions at runtime [configure Default Interactions runtime](/docs/mobile-monitoring/new-relic-mobile/mobile-sdk/configure-settings) and buildtimee [configure Default Interactions at buildtime](/docs/mobile-monitoring/new-relic-mobile-android/install-configure/configure-new-relic-gradle-plugin).

Reviewer: ```suggestion
  This feature works correctly only if [`Default Interactions`](/docs/mobile-monitoring/new-relic-mobile/mobile-sdk/configure-settings) are disabled at runtime and configured appropriately at [buildtime](/docs/mobile-monitoring/new-relic-mobile-android/install-configure/configure-new-relic-gradle-plugin). 
```

</Callout>

> # Creating Custom Interactions

Reviewer: ```suggestion
## Creating custom interactions
```

> You can create Custom Interactions and Enhance them with additional information. If Custom Interactions are not closed explicitly, New Relic agent will automatically close them and send the data to New Relic Platform.

Reviewer: ```suggestion
You can create custom interactions and enhance them with additional information. If custom interactions are not closed explicitly, the New Relic agent automatically closes them and sends the data to the New Relic Platform.
```

> To Create a Custom Interaction, use `startInteraction` to begin and `endInteraction` to close the interaction. The system automatically measures the timing.

Reviewer: ```suggestion
To create a custom interaction, use `startInteraction` to begin the interaction and `endInteraction` to close it. The system automatically measures the timing.
```

### Java

```java
// Start a custom interaction
String id = NewRelic.startInteraction("Tap on Search");

// ...do some work here...

// End the custom interaction
NewRelic.endInteraction(id);
```

### Kotlin

```kotlin
// Start a custom interaction
val id = NewRelic.startInteraction("Tap on Search")

// ...do some work here...

// End the custom interaction
NewRelic.endInteraction(id)
```

These methods allow you to capture the duration and details of specific interactions within your application, providing deeper insights into user behavior and application performance.

> # Creating Child Traces with Custom Interactions

Reviewer: ```suggestion
## Create child traces with custom interactions
```

To generate child traces, use the `NewRelic.startMethodTrace()` method. Here's how you can implement parent custom interactions and child traces:

### Java

```java
// Start a parent custom interaction
String parentId = NewRelic.startInteraction("Main Activity");

// Start a child trace
NewRelic.startMethodTrace("Load Resource From Database");

// ...do some work here...

// End the child trace
NewRelic.endMethodTrace();

// Start another child trace
NewRelic.startMethodTrace("Load Resource From Server");

// ...do some work here...

// End the child trace
NewRelic.endMethodTrace();

// End the parent interaction
NewRelic.endInteraction(parentId);
```

### Kotlin

```kotlin
// Start a parent custom interaction
val parentId = NewRelic.startInteraction("Main Activity")

// Start a child trace
NewRelic.startMethodTrace("Loop 1 Run")

// ...do some work here...

// End the child trace
NewRelic.endMethodTrace()

// Start another child trace
NewRelic.startMethodTrace("Loop 2 Run")

// ...do some work here...

// End the child trace
NewRelic.endMethodTrace()

// End the parent interaction
NewRelic.endInteraction(parentId)
```

## Considerations

> - If customers want to create custom interactions with method traces, they need to start and end the interaction without any user intervention.

Reviewer: ```suggestion
- If you want to create custom interactions with method traces, you need to start and end the interaction without any user intervention.
```

> - If customers wish to calculate the time between two interactions with user intervention, they should not create child traces for these interactions.

Reviewer: ```suggestion
- If you want to calculate the time between two interactions that involve user intervention, you should not create child traces for these interactions.
```

> This Approach allows for detailed tracking and measurement of interactions within your mobile application, providing valuable insights into application performance and user behavior.

Reviewer: ```suggestion
This approach enables detailed tracking and measurement of interactions within your mobile application to provide valuable insights into application performance and user behavior.
```
